rc/lsp.kak

```diff
@@ -2123,18 +2153,15 @@ decl -hidden int-list lsp_snippets_placeholder_groups
 face global SnippetsNextPlaceholders black,green+F
 face global SnippetsOtherPlaceholders black,yellow+F
 
-# First param is the text that was inserted in the completion, which will be deleted
-# Second param is the actual snippet
-def -hidden lsp-snippets-insert-completion -params 2 %{ eval -save-regs "a" %{
-    reg 'a' "%arg{1}"
-    exec -draft "<a-;><a-/><c-r>a<ret>d"
-    eval -draft -verbatim lsp-snippets-insert %arg{2}
-    remove-hooks window lsp-post-completion
-    hook -once -group lsp-post-completion window InsertCompletionHide .* %{
-        try %{
-            lsp-snippets-select-next-placeholders
-            exec '<a-;>d'
-        }
+declare-option -hidden str lsp_snippet_to_insert ""
+def -hidden lsp-snippets-insert-completion -params 1 %{ eval %{
+    set window lsp_snippet_to_insert %arg{1}
+    lsp-completion-on-dismiss %{
+        # Delete the inserted text
+        select %opt{lsp_completion_selection}
+        exec '<a-;>d'
```

Comment:
Given

	fn somefunction() {}
	fn main() {
	}

if I type `somefun<c-n>;<ret>` the semicolon will be erased.
I think that's pretty annoying. Sadly that's the range given to us by the InsertCompletionHide hook.
Maybe we can change the hook parameter. It should not include the semicolon.

Reply:
It's fairly easy to patch Kakoune to not include the last inserted key in the range specified by the hook parameter..
but it's probably not all that useful. How about instead we just replay the last key (that closed the completion pager) *if it's likely that it inserted something? This should definitely be a separate commit, doesn't need to be here. Roughly like below except we should probably reset to the empty string immediately after removing the `lsp-completion-dismissed` hooks

```
diff --git a/rc/lsp.kak b/rc/lsp.kak
index 76c69a49..8beea37f 100644
--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -367,8 +367,13 @@ define-command -hidden lsp-completion-dismissed -docstring "Called when the comp
 # User: lsp-completion-dismissed <command>
 # Run <command> when the completion menu is closed.
 # The inserted range is available as %opt{lsp_completion_selection} in the format expected by `select`.
+declare-option str lsp_key_that_dismissed_completion
 define-command -hidden lsp-completion-on-dismiss -params 1 %{
     hook -once -group lsp-completion-dismissed window User "^LspCompletionDismissed$" %arg{1}
+    set-option window lsp_key_that_dismissed_completion ""
+    hook -group lsp-completion-dismissed window RawKey [^<].*|<backspace>|<del>|<gt>|<lt>|<minus>|<percent>|<plus>|<right>|<semicolon>|<space>|<tab> %{
+        set-option window lsp_key_that_dismissed_completion %val{hook_param}
+    }
 }
 
 # Is called when a completion item is selected
@@ -381,6 +386,7 @@ define-command -hidden lsp-completion-item-resolve-request -params 1 \
     -docstring "Request additional attributes for a completion item" %{
     nop %sh{
         index=$kak_opt_lsp_completions_selected_item
+        exit
         [ "${index}" -eq -1 ] && exit
 
         (printf %s "
@@ -2164,6 +2170,7 @@ def -hidden lsp-snippets-insert-completion -params 1 %{ eval %{
         exec '<a-;>d'
         eval -draft -verbatim lsp-snippets-insert %opt[lsp_snippet_to_insert]
         try lsp-snippets-select-next-placeholders
+        exec -with-hooks %opt{lsp_key_that_dismissed_completion}
     }
 }}
```

Reply:
this is fixed by the above, right?

Reply:
yeah it means I can type `somefun<c-n>;<ret>` to get `somefunction();`, which is correct behavior I think. I'll test some more and submit it eventually unless you do

Reply:
Oh right, i understand the issue now. It might be worth looking at patching kakoune here, i cant exactly tell what the implications of that would be. Otherwise your solution above could work, but its the kind of tomfoolery i would really like to avoid

Reply:
I think I finally (after like 2 years) realized what needs to happen

When the completion pager is active, a key like 'a' "accepts" the completion.
After the completion is accepted, InsertCompletionHide hooks run.
What we actually want is that InsertCompletionHide runs before 'a' is processed.
That'd allow us to insert snippets, select the first placeholder and have 'a' be inserted there.
Of course this would mean that `<ret>` is usually not what the user wants to accept a snippet, but that's how it should be. Kakoune completions are accepted implicitly once the user continues typing.

So we just need to hack Kakoune to run InsertCompletionHide earlier. We're probably the only users who'd notice this change.

Reply:
submitted as https://github.com/mawww/kakoune/pull/4632
people who used `<ret>` to accept snippet completions would need to change habits, or should we special case `<ret>`? :thinking: